feat: add caller push of binary data

tomdcc opened the pull request:
Allows a caller to push binary data to an output stream rather than
having to provide a byte array or input stream. This gives the caller
more control over buffering strategy and allows explicit cleanup of
off-heap buffers or other non-garbage-collected resources.

Originally discussed on-list [here](https://www.postgresql.org/message-id/C659D6A4-430F-4F55-BE06-BE1C960A5405%40tomd.cc).

tomdcc commented:
FWIW the only failing test now is [unrelated to these changes](https://travis-ci.org/pgjdbc/pgjdbc/jobs/278012054#L5691)

vlsi commented on pgjdbc/src/main/java/org/postgresql/util/ByteBufferByteStreamWriter.java:
```diff
@@ -0,0 +1,49 @@
+  @Override
+  public void writeTo(OutputStream stream) throws IOException {
+    // this _does_ involve some copying to a temporary buffer, but that's unavoidable
+    // as OutputStream itself only accepts single bytes or heap allocated byte arrays
```
Does that mean we should focus on `transferTo(WritableByteChannel target)` kind of API instead of sticking with `OutputStream?

tomdcc replied:
I considered just trying to support someone setting a `ByteBuffer` as a parameter. But it has some issues.

Firstly, our internal socket isn't an NIO-based socket AFAICS, so we don't have a native writable channel to pass a byte buffer to. So we'd be reduced to doing what this class does and use a Channels wrapper, but if you look at the implementation it's just copying bytes into an intermediate buffer, exactly the same as we'd be doing if someone passes an InputStream. See comment further down for why that's not ideal.

I didn't want to go so far as to change our underlying socket implementation, but if / when we do that I think we could just accept a `ByteBuffer` as a parameter directly and then probably this interface would just be legacy.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/util/ByteStreamWriterTest.java:
```diff
@@ -0,0 +1,215 @@
+  private void insertStream(ByteStreamWriter writer) throws Exception {
+    PreparedStatement updatePS = con.prepareStatement(TestUtil.insertSQL("images", "img", "?"));
+    try {
+      updatePS.setObject(1, writer);
```
Technically speaking, JDBC has a standard API for this: `java.sql.PreparedStatement#setBinaryStream(int, java.io.InputStream, long)`

Have you considered something like `implements InputStream, ByteStreamWriter`?

tomdcc replied:
The issue with this interface is that it gives the caller no control over buffering strategy and makes cleanup harder. Passing in an `InputStream` means that the driver will always copy the bytes to a fixed-length buffer( currently 8k) and then copy them out, see `PgStream.sendStream()`. The caller then has to infer that it's safe to clean up the underlying resource by waiting for the expected number of bytes to be read.

In contrast the writer interface lets the caller write data straight to the BufferedOutputStream that wraps the socket. If it's writing more data than the buffered stream has then that buffer will get flushed and the data will be copied over by the underlying native `write` implementation in `SocketOutputStream` in a single go.

The caller also knows when the write has finished (since they're in control) and can clean up any off-heap resources immediately.

tomdcc replied:
Re-reading, I think that you're suggesting that the `ByteStreamWriter` should also implement `InputStream` as a means to pass it in. Is that right? The thing is, it _isn't_ an `InputStream` and the JDBC API has a perfectly valid method to pass proprietary objects through: `setObject()` :). Can you explain why you don't like that approach?

vlsi commented:
@tomdcc Does it make sense to have an abstraction over `OutputStream` for `writeTo(OutputStream stream)` API in order to have control over that and evolve properly?

tomdcc commented:
@vlsi What would that look like? If it's not an OutputStream then it'll have to have a very similar interface, won't it?

Or are you suggesting that we could use that interface to allow writing to other channel types in the future, e.g.:
```java
interface ByteTarget {
  // current implementation
  OutputStream getOutputStream();

  // some future implementation when we've got a NIO socket to use
  GatheringByteChannel getByteChannel();
}

interface public interface ByteStreamWriter {
  int getLength();
  void writeTo(ByteTarget target) throws IOException;
}
```

davecramer commented:
@tomdcc the challenge is that once we publish a public API we have to live with it for the foreseeable future. By abstracting it we can future proof it a bit hopefully.

tomdcc commented:
@davecramer Understood, thanks. @vlsi Is the above interface the kind of thing that you were thinking of?

tomdcc commented:
@davecramer @vlsi ping

davecramer commented:
@tomdcc I was kind of hoping more people would comment on this. Alas this doesn't seem to happen.
I'm OK with this interface.

@vlsi

tomdcc commented:
Sorry this sat for so long as our internal priorities shifted and I never got to giving it more time. I've now rebased it and introduced the above interface, as discussed.

Ping @vlsi @davecramer

paplorinc commented:
Rebased in https://github.com/pgjdbc/pgjdbc/pull/1659

paplorinc commented:
@tomdcc, merged in https://github.com/pgjdbc/pgjdbc/pull/1659#event-2909707244

davecramer commented:
closed with #1659
